### PR TITLE
Data cookbooks update 3

### PR DESCRIPTION
**Pull request by Jkovarik**

### CODE REVIEW

File: docs/data-cookbooks/error-handling.md [hunk 1]
```diff
@@ -17,100 +17,222 @@ The former (tasks which throw errors which are not `WorkflowError`s) is the expe
...
+* Tasks downstream of failed tasks should understand how to pass along exceptions if required.
+  If a task throws an error which is caught by the workflow configuration and passed to another state which also uses the CMA, the CMA overrides the exception key to `"None"` so the exception will not be passed to downstream tasks after the next state. This works if the exception is not needed in downstream tasks. If the exception is needed in downstream tasks, you need to re-attach the exception to the Cumulus message by setting the `ResultPath` to be `$.exception` for the task where the error is initially caught. In the example below, `CnmResponseFail` catches and re-attaches the error to the message.
```
Comment (markdboyd) [inline docs/data-cookbooks/error-handling.md]: Is there supposed to be an asterisk in front of this paragraph? To make it part of the same list as the other items below it

File: docs/data-cookbooks/hello-world.md [hunk 2]
```diff
@@ -12,50 +12,52 @@ Example task meant to be a sanity check/introduction to the Cumulus workflows.
...
 
-```yaml
-HelloWorld:
-  handler: index.handler
-  timeout: 300
-  memory: 256
-  source: 'node_modules/@cumulus/hello-world/dist/'
-  useMessageAdapter: true
-```
+If manually adding the resource to a deployment, the lambda resource source is located in the Cumulus source at  [`cumulus/tf-modules/ingest/hello-world-task.tf`](https://github.com/nasa/cumulus/tf-modules/ingest/hello-world-task.tf).  The lambda source code is located in the Cumulus source at ['cumulus/tasks/hello-world'](https://github.com/nasa/cumulus/tasks/hello-world/)
```
Comment (markdboyd) [inline docs/data-cookbooks/hello-world.md]:
```suggestion
If manually adding the resource to a deployment, the lambda resource source is located in the Cumulus source at  [`cumulus/tf-modules/ingest/hello-world-task.tf`](https://github.com/nasa/cumulus/tf-modules/ingest/hello-world-task.tf).  The lambda source code is located in the Cumulus source at ['cumulus/tasks/hello-world'](https://github.com/nasa/cumulus/tasks/hello-world/).
```

Comment (markdboyd) [inline docs/data-cookbooks/hello-world.md]: Could someone actually use the hello world task resource out of the cumulus module?

File: docs/data-cookbooks/run-tasks-in-lambda-or-docker.md [hunk 3]
```diff
@@ -33,75 +33,222 @@ If your task requires more than any of these resources or an unsupported runtime
...
-```yaml
-QueueGranules:
-  handler: index.handler
-  timeout: 900
-  memory: 3008
-  source: node_modules/@cumulus/queue-granules/dist/
-  useMessageAdapter: true
+# Example: Replacing AWS Lambda with a Docker container run on ECS
+
+This example will use an already-defined workflow from the `cumulus` module that includes the [`QueueGranules`] task(https://github.com/nasa/cumulus/blob/master/tf-modules/ingest/queue-granules-task.tf) in it's configuration.
```
Comment (markdboyd) [inline docs/data-cookbooks/run-tasks-in-lambda-or-docker.md]:
```suggestion
This example will use an already-defined workflow from the `cumulus` module that includes the [`QueueGranules` task](https://github.com/nasa/cumulus/blob/master/tf-modules/ingest/queue-granules-task.tf) in it's configuration.
```

File: docs/data-cookbooks/run-tasks-in-lambda-or-docker.md [hunk 4]
```diff
@@ -33,75 +33,222 @@ If your task requires more than any of these resources or an unsupported runtime
...
+
+The following example is an excerpt from the [Discover Granules workflow](https://github.com/nasa/cumulus/blob/master/tf-modules/ingest/queue-granules-task.tf) containing the step definition for the `Queue Granules` step:
```
Comment (markdboyd) [inline docs/data-cookbooks/run-tasks-in-lambda-or-docker.md]:
```suggestion
The following example is an excerpt from the [Discover Granules workflow](https://github.com/nasa/cumulus/blob/master/tf-modules/ingest/queue-granules-task.tf) containing the step definition for the `QueueGranules` step:
```

File: docs/data-cookbooks/run-tasks-in-lambda-or-docker.md [hunk 5]
```diff
@@ -33,75 +33,222 @@ If your task requires more than any of these resources or an unsupported runtime
...
+  }
+  command = [
+    "cumulus-ecs-task",
+    "--activityArn",
+    aws_sfn_activity.queue_granules.id,
+    "--lambdaArn",
+    module.cumulus.queue_granules_task_lambda_function_arn
+  ]
+  alarms = {
+    TaskCountHight = {
```
Comment (markdboyd) [inline docs/data-cookbooks/run-tasks-in-lambda-or-docker.md]:
```suggestion
    TaskCountHigh = {
```

Not your typo, but worth fixing

File: docs/data-cookbooks/run-tasks-in-lambda-or-docker.md [hunk 6]
```diff
@@ -33,75 +33,222 @@ If your task requires more than any of these resources or an unsupported runtime
...
+}
 ```
 
-and then modifying the corresponding Step Function State Machine definition in a workflow YAML file:
 
-```yaml
-    QueueGranules:
-      Type: Task
-      Resource: ${QueueGranulesActivity}
+* A instance of the `cumulus_ecs_service` module (found on the Cumulus [release](https://github.com/nasa/cumulus/releases page) configured to provide the QueueGranules task:
```
Comment (markdboyd) [inline docs/data-cookbooks/run-tasks-in-lambda-or-docker.md]:
```suggestion
* An instance of the `cumulus_ecs_service` module (found on the Cumulus [release](https://github.com/nasa/cumulus/releases page) configured to provide the QueueGranules task:
```

File: docs/data-cookbooks/run-tasks-in-lambda-or-docker.md [hunk 7]
```diff
@@ -33,75 +33,222 @@ If your task requires more than any of these resources or an unsupported runtime
...
+      metric_name         = "MemoryUtilization"
+      statistic           = "SampleCount"
+      threshold           = 1
+    }
+  }
+}
 ```
 
-## How do I name my resources? A short primer on Cumulus resource naming conventions
+* An updated [Discover Granules workflow](https://github.com/nasa/cumulus/blob/master/tf-modules/ingest/queue-granules-task.tf) to utilize the new resource (the resource key in the `QueueGranules` step has been updated to:
```
Comment (markdboyd) [inline docs/data-cookbooks/run-tasks-in-lambda-or-docker.md]: Is this the URL you're intending to link here? Seems like you mean to link to the definition of the discover granules workflow in our example?

File: docs/data-cookbooks/hello-world.md (hunk 2 above)
Comment (Jkovarik) [reply docs/data-cookbooks/hello-world.md]: I believe they could add the sub-module directly, same as we add the top module.

File: docs/data-cookbooks/run-tasks-in-lambda-or-docker.md [hunk 8]
```diff
@@ -33,75 +33,222 @@ If your task requires more than any of these resources or an unsupported runtime
...
+      "Cause": "Workflow failed"
+    }
+  }
+}
+JSON
+}
 
-Cumulus deployments currently depend on the [`kes`](https://github.com/developmentseed/kes) cli tool. `kes` references values in `app/config.yml`, `iam/config.yml`, lambda and workflow `.yml` files to populate the cloudformation templates that are a part of the `@cumulus/deployment` package. `kes` generates final cloudformation files and uploads them to AWS Cloudformation, creating or updating AWS Cloudformation Stacks.
-
-When deploying AWS Lambdas and AWS Activities as detailed above, note the following naming conventions:
```
Comment (markdboyd) [inline docs/data-cookbooks/run-tasks-in-lambda-or-docker.md]: Oh god I love to see this all deleted

File: docs/data-cookbooks/hello-world.md [hunk 9]
```diff
@@ -12,50 +12,52 @@ Example task meant to be a sanity check/introduction to the Cumulus workflows.
...
+}
 ```
 
+
 Workflow **error-handling** can be configured as discussed in the [Error-Handling](error-handling.md) cookbook.
 
 ### Task Configuration
 
-The HelloWorld [task itself](workflows/developing-workflow-tasks.md) is defined in [`cumulus/example/lambdas.yml`](https://github.com/nasa/cumulus/blob/master/example/lambdas.yml) under `HelloWorld:`
+The HelloWorld [task](workflows/developing-workflow-tasks.md) is provided for you as part of the `cumulus` terraform module, no configuration is needed.
```
Comment (markdboyd) [inline docs/data-cookbooks/hello-world.md]: Are you intending to link to the "developing tasks" doc or the hello world task definition?

File: docs/data-cookbooks/hello-world.md (hunk 2 above)
Comment (markdboyd) [reply docs/data-cookbooks/hello-world.md]: What sub-module? So their `source` would be `http://github.com/cumulus.zip//tf-modules/ingest`? How could they get only this task?

I'm just not sure what we mean by "manually adding the resource to a deployment". Do you mean the Cumulus module?

Comment (Jkovarik) [reply docs/data-cookbooks/hello-world.md]: Sorry, that wasn't stated quite right. 

What I mean is if they wanted the task (say to modify it) for experimentation reasons, they could *totally* grab that file and add it to their (custom module/directory/whatever) with a built lambda and add the lambda to a deployment with updates/whatever.        Perhaps .... we could phrase that a bit differently.   🤔

Comment (Jkovarik) [reply docs/data-cookbooks/hello-world.md]: They'd still need *ingest* to run a cumulus workflow for this example.

File: docs/data-cookbooks/run-tasks-in-lambda-or-docker.md (hunk 7 above)
Comment (Jkovarik) [reply docs/data-cookbooks/run-tasks-in-lambda-or-docker.md]: Not at all. 

79640e732

Comment (Jkovarik) [reply docs/data-cookbooks/run-tasks-in-lambda-or-docker.md]: No, it's not. 

Fixed: f7142c038

File: docs/data-cookbooks/hello-world.md (hunk 9 above)
Comment (Jkovarik) [reply docs/data-cookbooks/hello-world.md]: That was the *original* intent of it.  I just got rid of the redundant 'itself' here.      

I don't know that the task definition is particularly more or less helpful, but I can remove the link.

Comment (Jkovarik) [reply docs/data-cookbooks/hello-world.md]: 34b46e815

**Comment by Jkovarik** [comment]
@markdboyd  Thanks for the review.   Just one outstanding thing to discuss, else I think I've resolved all of the comments.

**Pull request merged**